apps/api/src/notifications/notifications.service.ts

```diff
@@ -0,0 +1,314 @@
+import { BadRequestException, Injectable, Logger } from '@nestjs/common'
+import { PrismaService } from '../prisma/prisma.service'
+import { SendConfirmationDto, SubscribePublicDto } from './dto/subscribe.dto'
+import { PersonService } from '../person/person.service'
+import { ConfirmConsentEmailDto } from '../email/template.interface'
+import { ConfigService } from '@nestjs/config'
+import { EmailService } from '../email/email.service'
+import {
+  EmailSentRegistry,
+  EmailType,
+  Person,
+  UnregisteredNotificationConsent,
+} from '@prisma/client'
+import { NotificationsProviderInterface } from './providers/notifications.interface.providers'
+import { SendGridParams } from './providers/notifications.sendgrid.types'
+import { DateTime } from 'luxon'
+import * as crypto from 'crypto'
+import { CampaignService } from '../campaign/campaign.service'
+import { KeycloakTokenParsed } from '../auth/keycloak'
+
+@Injectable()
+export class MarketingNotificationsService {
+  constructor(
+    private readonly prisma: PrismaService,
+    private readonly personService: PersonService,
+    private readonly config: ConfigService,
+    private readonly sendEmail: EmailService,
+    private readonly campaignService: CampaignService,
+    private readonly marketingNotificationsProvider: NotificationsProviderInterface<SendGridParams>,
+  ) {}
+
+  async sendConfirmConsentEmail(data: SendConfirmationDto) {
+    // Check if user is registered
+    let registered: Person | null
+    try {
+      registered = await this.personService.findByEmail(data.email)
+    } catch (e) {
+      return new BadRequestException('Failed to get email data')
+    }
+    // If registered and has given consent -> user is already subscribed
+    if (registered?.newsletter) return { message: 'Subscribed' }
+
+    // Check if un-registered consent already exists for this email
+    let unregisteredConsent: UnregisteredNotificationConsent | null
+    try {
+      unregisteredConsent = await this.prisma.unregisteredNotificationConsent.findFirst({
+        where: { email: data.email, consent: true },
+      })
+    } catch (e) {
+      return new BadRequestException('Failed to get email data')
+    }
+
+    // If unregistered consent exists -> user is already subscribed
+    if (unregisteredConsent?.consent) return { message: 'Subscribed' }
+
+    // Send confirmation email
+    try {
+      const sent = await this.prisma.emailSentRegistry.findFirst({
+        where: { email: data.email, type: EmailType.confirmConsent },
+      })
+      // Check if email was sent already in the past 1 minutes
+      const wasSent = this.emailWasSentRecently(sent, 1)
+
+      if (wasSent) return { message: 'Email Sent' }
```

Comment:
Good call to shield the mail sending!
I would make this even 10mins to avoid someone depleting our mail quota

Reply:
I'll put that to 10 minutes